Doc/library/stdtypes.rst

```diff
@@ -4822,33 +4822,53 @@ Generic Alias Type
+``GenericAlias`` objects are generally created by
+:ref:`subscripting<subscriptions>` a class. They are most often used with
+:ref:`container classes<sequence-types>`, such as :class:`list` or
+:class:`dict`. For example, ``list[int]`` is a ``GenericAlias`` object created
+by subscripting the :class:`list` class with the argument :class:`int`.
+``GenericAlias`` objects are intended primarily for use with
 :term:`type annotations <annotation>`.
 
-Usually, the :ref:`subscription <subscriptions>` of container objects calls the
-method :meth:`__getitem__` of the object.  However, the subscription of some
-containers' classes may call the classmethod :meth:`__class_getitem__` of the
-class instead. The classmethod :meth:`__class_getitem__` should return a
-``GenericAlias`` object.
-
 .. note::
-   If the :meth:`__getitem__` of the class' metaclass is present, it will take
-   precedence over the :meth:`__class_getitem__` defined in the class (see
-   :pep:`560` for more details).
 
-The ``GenericAlias`` object acts as a proxy for :term:`generic types
-<generic type>`, implementing *parameterized generics* - a specific instance
-of a generic which provides the types for container elements.
+   It is generally only possible to subscript a class if the class implements
+   the special method :meth:`~object.__class_getitem__`.
+
+A ``GenericAlias`` object acts as a proxy for a :term:`generic type`,
+implementing *parameterized generics*.
+
+For a container class, the
+argument(s) supplied to a :ref:`subscription<subscriptions>` of the class may
+indicate the type(s) of the elements an object contains. For example,
+``set[bytes]`` can be used in type annotations to signify a :class:`set` in
+which all the elements are of type :class:`bytes`.
+
+For a class which defines :meth:`~object.__class_getitem__` but is not a
+container, the argument(s) supplied to a subscription of the class will often
+indicate the return type(s) of one or more methods defined on an object. For
+example, :mod:`regular expressions<re>` can be used on both the :class:`str` data
+type and the :class:`bytes` data type:
+
+* If ``x = re.search('foo', 'foo')``, ``x`` will be a
+  :ref:`re.Match <match-objects>` object where the return values of
+  ``x.group(0)`` and ``x[0]`` will both be of type :class:`str`. We can
+  represent this kind of object in type annotations with the ``GenericAlias``
+  ``re.Match[str]``
 
-The user-exposed type for the ``GenericAlias`` object can be accessed from
-:class:`types.GenericAlias` and used for :func:`isinstance` checks.  It can
-also be used to create ``GenericAlias`` objects directly.
+* If ``y = re.search(b'bar', b'bar')``, however, ``y`` will also be an instance
+  of ``re.Match``, but the return values of ``y.group(0)`` and ``y[0]`` will
+  both be of type :class:`bytes`. In type annotations, we would represent this
+  subtype of :ref:`re.Match <match-objects>` with ``re.Match[bytes]``.
```

Comment:
I'm going to be pedantic here: is ``re.Match[bytes]`` really a subtype of ``re.Match[Any]``? I don't remember but I'm inclined to say no.

Reply:
I _think_ this is the correct terminology. But I'm not 100% sure, and now you mention it, I think it might be confusing wording for people who are new to type theory (sounds a bit too much like "subclass"). Maybe if we used the word "flavour" instead?

Reply:
Or "variety"?

Reply:
I think we can reuse the terminology in the previous paragraph, variety works good too. Good point about refraining from too much typing terminology in the stdtypes docs.

Also, I thought `re.Match` is invariant, so `re.Match[bytes]` shouldn't be a subtype of `re.Match[Any]`?

Reply:
Confession: I've never had any formal education in type theory (I did a music degree 😀). I was going off the definition of "subtype" [here](https://en.wikipedia.org/wiki/Subtyping):

> If S is a subtype of T, the subtyping relation is often written S <: T, to mean that any term of type S can be safely used in a context where a term of type T is expected.

mypy [is happy](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=6090ffeeab4f7b0167f25ae62488761b) with us using `Match[bytes]` in any situation where `Match[Any]` is expected -- and that seems correct, to me?

But regardless, we should use different language here, we agree 😀

Reply:
> Confession: I've never had any formal education in type theory (I did a music degree 😀). I was going off the definition of "subtype" here:

Well same here (not the music degree part though), I have no clue what I'm doing most of the time 😆.

Also oh, it seems that `typing.Match` is special, it's [generic via `typing.Match[typing.AnyStr]`](https://docs.python.org/3/library/typing.html#typing.Match), so I don't think its variance matters at all and I was probably wrong.

Thanks for the change.